crates/core/src/checkpoint.rs

```diff
@@ -192,7 +163,57 @@ impl CheckPoint {
             })
     }
 
-    /// Inserts `block_id` at its height within the chain.
+    /// This method tests for `self` and `other` to have equal internal pointers.
+    pub fn eq_ptr(&self, other: &Self) -> bool {
+        Arc::as_ptr(&self.0) == Arc::as_ptr(&other.0)
+    }
+}
+
+// Methods where `D: ToBlockHash`
+impl<D> CheckPoint<D>
+where
+    D: ToBlockHash + fmt::Debug + Copy,
+{
+    /// Construct a new base [`CheckPoint`] from given `height` and `data` at the front of a linked
+    /// list.
+    pub fn new(height: u32, data: D) -> Self {
+        Self(Arc::new(CPInner {
+            block_id: BlockId {
+                height,
+                hash: data.to_blockhash(),
+            },
+            data,
+            prev: None,
+        }))
+    }
+
+    /// Construct from an iterator of block data.
+    ///
+    /// Returns `Err(None)` if `blocks` doesn't yield any data. If the blocks are not in ascending
+    /// height order, then returns an `Err(..)` containing the last checkpoint that would have been
+    /// extended.
+    pub fn from_blocks(blocks: impl IntoIterator<Item = (u32, D)>) -> Result<Self, Option<Self>> {
+        let mut blocks = blocks.into_iter();
+        let (height, data) = blocks.next().ok_or(None)?;
+        let mut cp = CheckPoint::new(height, data);
+        cp = cp.extend(blocks)?;
+
+        Ok(cp)
+    }
+
+    /// Extends the checkpoint linked list by a iterator containing `height` and `data`.
+    ///
+    /// Returns an `Err(self)` if there is block which does not have a greater height than the
+    /// previous one.
+    pub fn extend(self, blockdata: impl IntoIterator<Item = (u32, D)>) -> Result<Self, Self> {
+        let mut cp = self.clone();
+        for (height, data) in blockdata {
+            cp = cp.push(height, data)?;
+        }
+        Ok(cp)
+    }
+
+    /// Inserts `data` at its `height` within the chain.
     ///
     /// The effect of `insert` depends on whether a height already exists. If it doesn't the
```

Comment:
nit: it's not expecting `BlockId` as a parameter anymore, although it's constructed internally for the checkpoint, I think we could update its references in the doc.